### PR TITLE
Enable the mono subset to build on FreeBSD x64

### PR DESCRIPTION
**Pull request by jasonpugsley**
Add the necessary CMake settings to detect the CPU and
reference the correct libraries when the build host is FreeBSD.

These changes allow me to build the mono subset on FreeBSD 12.2
There shouldn't be any impact when building on other hosts. I also did a cross-build
of FreeBSD on Linux and that completed successfully (it already builds correctly there).

https://github.com/dotnet/runtime/issues/14537#issuecomment-891336597

### CODE REVIEW

**Review by vargaz — APPROVED** [review]

**Comment by Thefrank** [comment]
`set(HAVE_ALLOCA_H 0)`
Does it think we have this when we don't?

**Comment by jasonpugsley** [comment]
> `set(HAVE_ALLOCA_H 0)`
> Does it think we have this when we don't?

I get a build error:
```
  mono/mini/../../mono/eglib/eglib-config.h:36:6: error: invalid token at start of a preprocessor expression
  #if  == 1
       ^
```
From here `artifacts/obj/mono/FreeBSD.x64.Debug/mono/eglib/eglib-config.h`
```C
#if  == 1
#define G_HAVE_ALLOCA_H
#endif
```

**Comment by Thefrank** [comment]
ah!
There is a note in src/mono/cmake/configure.cmake about detection not working :(

Speaking of... would it be better/cleaner to move:
```
  set(HAVE_SYS_ICU 1)
  set(HAVE_ALLOCA_H 0)
  set(HAVE_SYS_SYSCTL_H 1)
  set(HAVE_SYS_USER_H 1)
```
into its own `if(HOST_FREEBSD)` block at the end in src/mono/cmake/configure.cmake? The fix for Solaris that was recently committed put those `set` fixes in that spot.

the `add_link_options(-lpthread -lm)` would go into src/mono/mono/mini/CMakeLists.txt as
```
elseif(HOST_FREEBSD)
  set(OS_LIBS pthread m ${CMAKE_DL_LIBS})
```
after the SOLARIS elseif if following the same pattern. (might not need the `CMAKE_DL_LIBS`)

`--subset mono` builds either way.

**Comment by jasonpugsley** [comment]
OK this needs a little more discussion. It turns out all but one use of HAVE_ALLOCA_H are #ifdefs. So setting it to be 0 doesn't help there - any value satisfies an #ifdef. The only other use case is to set up G_HAVE_ALLOCA_H for just a single reference in `src/mono/mono/eglib/glib.h`
Ideally we would modify glib.h to change `G_HAVE_ALLOCA_H` to `HAVE_ALLOCA_H`.
It is currently
```C
#ifdef G_HAVE_ALLOCA_H
#include <alloca.h>
#endif
```
If that's not desirable, then can we change `src/mono/cmake/eglib-config.h.cmake.in`
So from
```
#if @HAVE_ALLOCA_H@ == 1
#define G_HAVE_ALLOCA_H
#endif
```
to
```
#ifdef HAVE_ALLOCA_H
#define G_HAVE_ALLOCA_H
#endif
```

**Comment by jasonpugsley** [comment]
Following up on my last question, if the variable is set to 0 in the cmake file  `set(HAVE_ALLOCA_H 0)`  then it is not defined in the .h.in -> .h generated file. But its value _can_ be referenced which is why the `%HAVE_ALLOCA_H% == 1` requires it to be set, but fails to work if it's not defined.
I guess FreeBSD must be the only host platform that doesn't have alloca.h
On a side note, when cross compiling on Linux, the alloca.h file is found. This would mean the host root files are also searched instead of only the cross root files.

**Comment by Thefrank** [comment]
The way that `#cmakedefine` works is the variable will either be set or an empty string :(

This is why FreeBSD gets that strange ` #if  == 1` error
Something like `#cmakedefine01` found in newer versions of CMake still ends with an empty string during my testing
(https://cmake.org/cmake/help/latest/command/configure_file.html)

**Comment by jasonpugsley** [comment]
In my testing I found this works fine on Linux (native), on FreeBSD (native) and on Linux (cross-building FreeBSD) for `src/mono/cmake/eglib-config.h.cmake.in`
```C
#ifdef HAVE_ALLOCA_H
#define G_HAVE_ALLOCA_H
#endif
```

**Comment by Thefrank** [comment]
At this point ill take what I can get. The CI system will let us know if anything explodes [it shouldn't].

**Comment by jasonpugsley** [comment]
I have refined the changes to be more _complete_. I can remove the changes to find-native-compiler.sh and init-compiler.sh if they shouldn't be done here - I have been using clang-12 on Linux and FreeBSD without issue.

I have made the change to `src/mono/cmake/eglib-config.h.cmake.in` as per the previous comments but please review this for consistency.

**Comment by jasonpugsley** [comment]
The cross-build in CI failed because the container is missing or can't find lld. We'll need to update some of the dependencies.

**Review by josteink — APPROVED** [review]
LGTM. Hope someone can get that container fixed.

**Comment by jasonpugsley** [comment]
This container
`ubuntu-18.04-cross-freebsd-12-20210813173449-f13d79e`
has an updated crossfootfs. It's FreeBSD12.1
I've updated `eng/pipelines/common/platform-matrix.yml`
Hopefully that works.

**Comment by Thefrank** [comment]
It does appear that the freebsd12 container is still on 12.1.
It looks like the changes from https://github.com/dotnet/arcade/pull/7588 haven't made their way into the containers yet.

File: eng/common/cross/toolchain.cmake [hunk 1]
```diff
@@ -90,6 +90,9 @@ elseif(CMAKE_SYSTEM_NAME STREQUAL "FreeBSD")
     set(CMAKE_C_COMPILER_TARGET ${triple})
     set(CMAKE_CXX_COMPILER_TARGET ${triple})
     set(CMAKE_ASM_COMPILER_TARGET ${triple})
+    set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -fuse-ld=lld")
```
Comment (arrowd) [inline eng/common/cross/toolchain.cmake:93]: Why is this required? The base compiler already uses `lld` by default. Is for cases when GCC from ports is being used as the compiler?

**Comment by jasonpugsley** [comment]
The 3 FreeBSD cross-builds in the checks succeeded but FreeBSD/mono isn't actually included in runtime.yml yet. That can go in another change set as it affects all build runs in the CI system.

File: eng/common/cross/toolchain.cmake (hunk 1 above)
Comment (jasonpugsley) [reply eng/common/cross/toolchain.cmake:93]: It seems on a Linux cross-build that it's defaulting to the system linker which wasn't creating completely valid FreeBSD objects. The clang LLD linker creates valid FreeBSD objects.

Comment (arrowd) [reply eng/common/cross/toolchain.cmake:93]: Ah, this stuff also works when cross-compiled, wow. The flag makes sense, then.

File: src/mono/cmake/configure.cmake [hunk 2]
```diff
@@ -175,3 +175,9 @@ if(HOST_SOLARIS)
   set(HAVE_NETINET_TCP_H 1)
   set(HAVE_GETADDRINFO 1)
 endif()
+
+if(CMAKE_SYSTEM_NAME STREQUAL "FreeBSD")
+  # FreeBSD fails earlier ac_check_headers for these because <sys/types.h> is needed.
```
Comment (wfurt) [inline src/mono/cmake/configure.cmake:180]: I'm wondering if we should fix the detection at some point..

**Review by wfurt — APPROVED** [review]
LGTM

File: src/mono/cmake/configure.cmake (hunk 2 above)
Comment (Thefrank) [reply src/mono/cmake/configure.cmake:180]: the `ac_check_headers` for the items that we must manually define will work fine for Linux systems which don't need `<sys/types.h>`.

Comment (wfurt) [reply src/mono/cmake/configure.cmake:180]: We could add them even for Linux, right? I think we did it in few other places. But I don't see it critical - just may be easier to maintain.

Comment (jasonpugsley) [reply src/mono/cmake/configure.cmake:180]: I tried a few different ways to do that but I'm new at CMake and couldn't get it working here without brute force. It's probably something simple I missed.

Comment (Thefrank) [reply src/mono/cmake/configure.cmake:180]: On the coreclr side, `sysctl` is explictly marked as depreciated on Linux via comment and `set(HAVE_SYS_SYSCTL_H 0)` via a `if (CLR_CMAKE_TARGET_LINUX)` (https://github.com/dotnet/runtime/blob/214443193a83b47170a8bec7ab4247ca745ccc8b/src/libraries/Native/Unix/configure.cmake#L904-L#911)

Comment (jasonpugsley) [reply src/mono/cmake/configure.cmake:180]: I'd need someone to suggest a change here - I don't have a different solution.

Comment (Thefrank) [reply src/mono/cmake/configure.cmake:180]: honestly? leave it as is. most of the `ac_check_headers` and `ac_check_funcs` should be reworked so that all these other platforms don't need to make their own hardcoded settings, but that might be quite a challenge. 

If you really want to change the detection you can do what was done in the `src/libraries/Native/Unix/configure.cmake` (linked above)

Comment (jasonpugsley) [reply src/mono/cmake/configure.cmake:180]: I'm not sure the solution in the link is any more simple/clean. I'm inclined to leave it for now and let someone with a stronger conviction take a go at it later.

**Review by vargaz — APPROVED** [review]

**Comment by jasonpugsley** [comment]
I noticed libinotify wasn't being found on the cross builds so I've fixed that now.

**Comment by danmoseley** [comment]
@jkotas you are more involved with efforts to bring up/maintain other platforms. does this seem a reasonable change to take at the moment? I see no problem if there is negligible risk to anything else.

**Comment by jkotas** [comment]
This is in the same category as https://github.com/dotnet/runtime/pull/58085#issuecomment-905212795 . There are likely more changes needed to make FreeBSD work well. It is unlikely we will need to be able to port them all to the release branch.

After thinking about it a bit, I would recommend to keep the policy simple and only backport changes for officially supported OSes and architectures to the release branches. If somebody wants to build OSes that are not officially supported by the .NET team, they can do it either from main or release branch w/ local patch. Tizen OS has been on this plan too.

**Comment by danmoseley** [comment]
I assume that includes changes to RIDs?

**Comment by Thefrank** [comment]
Where does that leave FreeBSD (and also illumos) for net6? 

Currently under net5: `runtime` crossbuilds Linux->FreeBSD with no patches. Both `aspnetcore` and `installer` need a `freebsd-x64` RID manually added in via `sed`/patch/diff but will also build without patches. 

I was recently able to make a usable net6-preview7 but it required 3 freebsd-based patches, or 5 total patches as two of the three freebsd ones require earlier illumos-based patches (cherry-pick did not work otherwise). While this is annoying, having the 3(5?) patches backported into net6 will make it much easier for the freebsd community to build/help support net6 during its lifetime. I understand if a backport will not be suitable here, but I can list all of the needed patches if it will help.

Finally, I would like to thank the dotnet team for allowing unofficial platforms/arches to be added into dotnet :)
